diff --git a/otherlibs/configurator/src/v1.ml b/otherlibs/configurator/src/v1.ml
index dffb7832b..50520de17 100644
--- a/otherlibs/configurator/src/v1.ml
+++ b/otherlibs/configurator/src/v1.ml
@@ -388,8 +388,7 @@ let create ?dest_dir ?ocamlc ?(log = ignore) name =
     in
     fill_in_fields_that_depends_on_ocamlc_config { t with ocamlc_config }
 
-let need_to_compile_and_link_separately t =
-  (* Vague memory from writing the discover.ml script for Lwt... *)
+let is_msvc t =
   match t.ccomp_type with
   | "msvc" -> true
   | _ -> false
@@ -399,7 +398,6 @@ let compile_and_link_c_prog t ?(c_flags = []) ?(link_flags = []) code =
   Unix.mkdir dir 0o777;
   let base = dir ^/ "test" in
   let c_fname = base ^ ".c" in
-  let obj_fname = base ^ t.ext_obj in
   let exe_fname = base ^ ".exe" in
   Io.write_file c_fname code;
   logf t "compiling c program:";
@@ -407,17 +405,21 @@ let compile_and_link_c_prog t ?(c_flags = []) ?(link_flags = []) code =
   let run_ok args =
     Process.run_command_ok t ~dir (Process.command_args t.c_compiler args)
   in
-  let ok =
-    if need_to_compile_and_link_separately t then
-      run_ok (c_flags @ [ "-I"; t.stdlib_dir; "-c"; c_fname ])
-      && run_ok ("-o" :: exe_fname :: obj_fname :: link_flags)
+  let output_flag =
+    if is_msvc t then
+      [ "-Fe" ^ exe_fname ]
     else
-      run_ok
-        (List.concat
-           [ c_flags
-           ; [ "-I"; t.stdlib_dir; "-o"; exe_fname; c_fname ]
-           ; link_flags
-           ])
+      [ "-o"; exe_fname ]
+  in
+  let ok =
+    run_ok
+      (List.concat
+         [ c_flags
+         ; [ "-I"; t.stdlib_dir ]
+         ; output_flag
+         ; [ c_fname ]
+         ; link_flags
+         ])
   in
   if ok then
     Ok ()
@@ -434,9 +436,16 @@ let compile_c_prog t ?(c_flags = []) code =
   logf t "compiling c program:";
   List.iter (String.split_lines code) ~f:(logf t " | %s");
   let ok =
+    let output_flag =
+      if is_msvc t then
+        [ "-Fo" ^ obj_fname ]
+      else
+        [ "-o"; obj_fname ]
+    in
     Process.run_command_ok t ~dir
       (Process.command_args t.c_compiler
-         (c_flags @ [ "-I"; t.stdlib_dir; "-o"; obj_fname; "-c"; c_fname ]))
+         (List.concat
+            [ c_flags; [ "-I"; t.stdlib_dir ]; output_flag; [ "-c"; c_fname ] ]))
   in
   if ok then
     Ok obj_fname
