diff --git a/configure b/configure
index 47754ef..c0a5dde 100755
--- a/configure
+++ b/configure
@@ -12843,7 +12843,11 @@ $as_echo "$as_me: WARNING: Consider using GCC version 4.2 or above." >&2;};
       internal_cflags="$cc_warnings -fno-common \
 -fexcess-precision=standard" ;; #(
   msvc-*) :
-    common_cflags="-nologo -O2 -Gy- -MD $cc_warnings"
+    if [ -n "${CFLAGS_MSVC}" ]; then
+      common_cflags="-nologo -Gy- ${CFLAGS_MSVC} $cc_warnings"
+    else
+      common_cflags="-nologo -O2 -Gy- -MD $cc_warnings"
+    fi
       common_cppflags="-D_CRT_SECURE_NO_DEPRECATE"
       internal_cppflags='-DUNICODE -D_UNICODE'
 
